endpoints-control/src/main/java/com/google/api/control/aggregator/QuotaRequestAggregator.java

```diff
@@ -0,0 +1,296 @@
+  private final ConcurrentLinkedDeque<AllocateQuotaRequest> out;
+  private final String serviceName;
+  private final Ticker ticker;
+  private final Cache<String, CachedItem> cache;
+  private final QuotaAggregationOptions options;
+  private final long timeoutIntervalNs;
+  private boolean inFlushAll;
+
+  public QuotaRequestAggregator(String serviceName, QuotaAggregationOptions options,
+      @Nullable Ticker ticker) {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(serviceName),
+        "service name cannot be empty");
+    Preconditions.checkNotNull(options, "options must be non-null");
+    this.out = new ConcurrentLinkedDeque<>();
+    this.ticker = ticker == null ? Ticker.systemTicker() : ticker;
+    this.serviceName = serviceName;
+    this.options = options;
+    this.cache = createCache(this.ticker);
+    this.timeoutIntervalNs = options.getTimeoutMillis() * NANOS_PER_MILLI;
+  }
+
+  /**
+   * @return the interval in milliseconds between calls to {@link #flush}
+   */
+  public int getFlushIntervalMillis() {
+    if (cache == null) {
+      return NON_CACHING;
+    } else {
+      return options.getRefreshMillis();
+    }
+  }
+
+  public List<AllocateQuotaRequest> flush() {
+    if (cache == null) {
+      return ImmutableList.of();
+    }
+
+    synchronized (cache) {
+      cache.cleanUp();
+      for (Map.Entry<String, CachedItem> entry : cache.asMap().entrySet()) {
+        CachedItem item = entry.getValue();
+        if (!inFlushAll && !shouldDrop(item)) {
+          if (!item.isInFlight && item.aggregator != null) {
+            item.isInFlight = true;
+            item.lastRefreshTimestamp = ticker.read();
+            out.add(item.extractRequest());
+          }
+        }
+      }
+      List<AllocateQuotaRequest> reqs = new ArrayList<>(out);
+      out.clear();
+      return reqs;
+    }
+  }
+
+  /**
+   * Clears this instances cache of aggregated operations.
+   *
+   * Is intended to be called by the driver before shutdown.
+   */
+  public void clear() {
+    if (cache == null) {
+      return;
+    }
+    synchronized (cache) {
+      inFlushAll = true;
+      cache.invalidateAll();
+      out.clear();
+      inFlushAll = false;
+    }
+  }
+
+  public AllocateQuotaResponse allocateQuota(AllocateQuotaRequest req) {
+    Preconditions.checkArgument(req.getServiceName().equals(serviceName), "service name mismatch");
+    Preconditions.checkArgument(
+        req.hasAllocateOperation(), "expected quota operation was not present");
+
+    if (cache == null) {
+      // By returning NO_FOUND, caller will send request to server.
+      return null;
+    }
+
+    String signature = sign(req).toString();
+    synchronized (cache) {
+      CachedItem item = cache.getIfPresent(signature);
+      if (item == null) {
+        // To avoid sending concurrent allocateQuota from concurrent requests.
+        // insert a temporary positive response to the cache. Requests from other
```

Comment:
"requests from other requests" is a phrase that takes a little time to work out what it means. Can you clarify?